api/saml/controller.py

```diff
@@ -0,0 +1,249 @@
+        :type authenticator: Authenticator
+
+        :param authentication_manager_factory: SAML authentication manager factory
+        :type authentication_manager_factory: SAMLAuthenticationManagerFactory
+        """
+        self._circulation_manager = circulation_manager
+        self._authenticator = authenticator
+        self._authentication_manager_factory = authentication_manager_factory
+
+    def _get_authentication_manager(self, db, authentication_provider):
+        """Returns an instance of SAML authentication manager
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :param authentication_provider: SAML authentication provider
+        :type authentication_provider: SAMLAuthenticationProvider
+
+        :return: Authentication manager
+        :rtype: SAMLAuthenticationManager
+        """
+        return self._authentication_manager_factory.create(authentication_provider.external_integration(db))
+
+    def _add_params_to_url(self, url, params):
+        """Adds parameters as a query part of the URL
+
+        :param url: URL
+        :type url: string
+
+        :param params: Dictionary containing parameters
+        :type params: Dict
+
+        :return: URL with parameters formatted as a query string
+        :rtype: string
+        """
+        query = urllib.urlencode(params)
+
+        if '?' in url:
+            url += '&' + query
+        else:
+            url += '?' + query
+
+        return url
+
+    def _error_uri(self, redirect_uri, problem_detail):
+        """Encodes the given ProblemDetail into the fragment identifier of the given URI
+
+        :param redirect_uri: Redirection URL
+        :type redirect_uri: string
+
+        :param problem_detail: ProblemDetail object describing the error
+        :type problem_detail: ProblemDetail
+
+        :return: Redirection URL
+        :rtype: string
+        """
+        problem_detail_json = pd_json(
+            problem_detail.uri,
+            problem_detail.status_code,
+            problem_detail.title,
+            problem_detail.detail,
+            problem_detail.instance,
+            problem_detail.debug_message
+        )
+        params = {
+            self.ERROR: problem_detail_json
+        }
+        redirect_uri = self._add_params_to_url(redirect_uri, params)
+
+        return redirect_uri
+
+    def _get_redirect_uri(self, relay_state):
+        """Returns a redirection URL from the relay state
+
+        :param relay_state: SAML response's relay state
+        :type relay_state: string
+
+        :return: Redirection URL
+        :rtype: string
+        """
+        relay_state_parse_result = urlparse.urlparse(relay_state)
+        relay_state_parameters = urlparse.parse_qs(relay_state_parse_result.query)
+
+        del relay_state_parameters[self.PROVIDER_NAME]
+        del relay_state_parameters[self.IDP_ENTITY_ID]
+        del relay_state_parameters[self.LIBRARY_SHORT_NAME]
```

Comment:
This will raise KeyError if for whatever reason the relay state URL doesn't include these keys. I believe this URL is generated by the circulation manager, so this shouldn't happen in normal usage. But we try to avoid uncaught exceptions even if they could only be triggered by bad input.

Reply:
Added additional checks whether a specific key in the dictionary